src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -3056,7 +3091,138 @@ public static void Main()
 ";
             var comp = CreateCompilation(source, options: TestOptions.DebugExe);
             comp.VerifyDiagnostics();
-            CompileAndVerify(comp, expectedOutput: "True");
+            var verifier = CompileAndVerify(comp, expectedOutput: "True");
+            verifier.VerifyIL("C.Main", @"{
+  // Code size       57 (0x39)
+  .maxstack  3
+  .locals init (System.ValueTuple<int, int> V_0,
+                System.ValueTuple<int, int> V_1)
+  IL_0000:  nop
+  IL_0001:  ldloca.s   V_0
+  IL_0003:  ldc.i4.1
+  IL_0004:  ldc.i4.2
+  IL_0005:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_000a:  ldloca.s   V_1
+  IL_000c:  ldc.i4.1
+  IL_000d:  ldc.i4.2
+  IL_000e:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_0013:  ldloc.0
+  IL_0014:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_0019:  ldloc.1
+  IL_001a:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_001f:  bne.un.s   IL_0031
+  IL_0021:  ldloc.0
+  IL_0022:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_0027:  ldloc.1
+  IL_0028:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_002d:  ceq
+  IL_002f:  br.s       IL_0032
+  IL_0031:  ldc.i4.0
+  IL_0032:  call       ""void System.Console.Write(bool)""
+  IL_0037:  nop
+  IL_0038:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_OneSideNeverNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write(((int, int)?) (1, 2) == (1, 2));
+    }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "True");
+            verifier.VerifyIL("C.Main", @"{
+  // Code size       38 (0x26)
+  .maxstack  3
+  .locals init (System.ValueTuple<int, int> V_0)
+  IL_0000:  nop
+  IL_0001:  ldloca.s   V_0
+  IL_0003:  ldc.i4.1
+  IL_0004:  ldc.i4.2
+  IL_0005:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_000a:  ldloc.0
+  IL_000b:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_0010:  ldc.i4.1
+  IL_0011:  bne.un.s   IL_001e
+  IL_0013:  ldloc.0
+  IL_0014:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_0019:  ldc.i4.2
+  IL_001a:  ceq
+  IL_001c:  br.s       IL_001f
+  IL_001e:  ldc.i4.0
+  IL_001f:  call       ""void System.Console.Write(bool)""
+  IL_0024:  nop
+  IL_0025:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_TupleAlwaysNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write(((int, int)?)null == ((int, int)?)null);
```

Comment:
>((int, int)?)null == ((int, int)?)null [](start = 29, length = 38)

It would be good to cover all combinations: left (might be null, never null, always null) vs. right (never null, always null). And switch sides too.